src/instruction.rs

```diff
@@ -0,0 +1,185 @@
+//! Program instructions
+
+use solana_program::clock::UnixTimestamp;
+
+/// Instructions supported by the staking program
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum StakeInstruction {
+    /// Creates Stake config account which controls staking parameters
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[]` Slash authority, may be a PDA for a governance program
+    /// 2. `[]` Config authority, may be a PDA for a governance program
+    /// 3. `[]` Stake Token Mint
+    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
+    ///   Must be empty and owned by PDA with seeds `['token-owner', stake_config]`
+    InitializeConfig {
+        cooldown_time: UnixTimestamp,
+        max_deactivation_basis_points: u16,
+    },
+
+    /// Initializes stake account data for a validator.
+    ///
+    /// NOTE: Anybody can create the stake account for a validator. The authority
+    /// defaults to the validator withdraw authority.
+    ///
+    /// 0. `[]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    ///     * PDA seeds: ['stake', validator, config_account]
+    /// 2. `[]` Validator vote account
+    /// 3. `[]` System program
+    InitializeStake,
+
+    /// Stakes tokens with the given config.
+    ///
+    /// Limited to the current amount of SOL staked to the validator.
+    ///
+    /// NOTE: Anybody can stake tokens to a validator, but this does not work
+    /// like native staking, because the validator can take control of staked
+    /// tokens by deactivating and withdrawing.
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    ///     * PDA seeds: ['stake', validator, config_account]
+    /// 2. `[writable]` Token Account
+    /// 3. `[signer]` Owner or delegate of the token account
+    /// 4. `[]` Validator vote account
+    /// 3. `[]` Stake Token Mint
+    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
+    ///   Must be the token account on the stake config account
+    /// 5. `[]` Token program
+    /// 6.. Extra accounts required for the transfer hook
```

Comment:
Ahh interesting. So just for my understanding, the act of staking PAL tokens is effectively a transfer, which triggers the transfer hook to the Rewards program?